service/history/replication/eventhandler/history_events_handler.go

```diff
@@ -0,0 +1,156 @@
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package eventhandler
+
+import (
+	"context"
+
+	historypb "go.temporal.io/api/history/v1"
+	"go.temporal.io/api/serviceerror"
+	historyspb "go.temporal.io/server/api/history/v1"
+	workflowpb "go.temporal.io/server/api/workflow/v1"
+	"go.temporal.io/server/common/definition"
+	"go.temporal.io/server/common/persistence/versionhistory"
+	"go.temporal.io/server/service/history/replication"
+)
+
+//go:generate mockgen -copyright_file ../../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination history_events_handler_mock.go
+
+// Past vs Future
+// Past and Future are introduced to handle the case:
+//
+//		wf_1 was originally started in cluster_A, and was migrated to cluster_B.
+//	 Then wf_1 data was deleted in cluster_A. Then we want to migrate the wf_1 data back to cluster_A.
+//	 in this case, past events are the event 1 to the last events that are generated by wf_1 (event_version % failover_incremental == cluster_A's initial_failover_version)
+type (
+	// HistoryEventsHandler is to handle all cases that add history events from remote cluster to current cluster.
+	// so it can be used by:
+	// 1. ExecutableHistoryTask to replicate events
+	// 2. When HistoryResender trying to add events (currently triggered by RetryReplication error, which still need to handle import case)
+	HistoryEventsHandler interface {
+		HandleHistoryEvents(
+			ctx context.Context,
+			sourceClusterName string,
+			workflowKey definition.WorkflowKey,
+			baseExecutionInfo *workflowpb.BaseExecutionInfo,
+			versionHistoryItems []*historyspb.VersionHistoryItem,
+			historyEvents [][]*historypb.HistoryEvent,
+			newEvents []*historypb.HistoryEvent,
+		) error
+	}
+
+	historyEventsHandlerImpl struct {
+		replication.ProcessToolBox
+		PastEventsHandler
```

Comment:
I feel the past/future is not good wording here. It is hard to understand. could it just be import/replicate? Or LocalEventHandler and RemoteEventHandler?

Reply:
Yeah, I like "LocalEventHandler and RemoteEventHandler". Will change it in next revision.

Reply:
I actually asked GPT to come up with a name, Local and Remote is one of them!